### PR TITLE
Add support for remote notifications

### PR DESCRIPTION
**Pull request by aaemnnosttv**
## Summary

Addresses issue #1110

## Relevant technical choices
- Proxy API responses are mocked during E2E
- Added `dompurify` to allow for limited HTML in content passed to notification `description`
- Increased bundle size limit by 10KB (see #1180)
- ~~Updated VRT due to visual changes (see https://github.com/google/site-kit-wp/pull/1170#issuecomment-591315512)~~

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: includes/Core/Notifications/Notifications.php [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+							$credentials = $this->credentials->get();
+							$response    = wp_remote_post(
+								$this->google_proxy->url( '/notifications/mark' ),
+								array(
+									'body' => array(
+										'site_id'         => $credentials['oauth2_client_id'],
+										'site_secret'     => $credentials['oauth2_client_secret'],
+										'notification_id' => $data['notificationID'],
+										'notification_state' => $data['notificationState'],
```
Comment (aaemnnosttv) [inline includes/Core/Notifications/Notifications.php]: @felixarntz I'm not sure why, but PHPCS says that this alignment is correct.

File: includes/Core/Notifications/Notifications.php [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+		$body    = wp_remote_retrieve_body( $response );
+		$decoded = json_decode( $body, true );
+
+		if ( json_last_error() ) {
+			throw new Exception( 'Error while decoding response: ' . json_last_error() );
+		}
+
+		if ( ! empty( $decoded['error'] ) ) {
+			throw new Exception( $decoded['error'] );
+		}
```
Comment (aaemnnosttv) [inline includes/Core/Notifications/Notifications.php]: @felixarntz Should we ever display an error message returned by the remote? Currently it's logged in the console only.

File: includes/Core/Notifications/Notifications.php [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+	 * @return WP_Error
+	 */
+	private function missing_required_param( $param ) {
+		return new WP_Error(
+			'missing_required_param',
+			/* translators: %s: Missing parameter name */
+			sprintf( __( 'Request parameter is empty: %s.', 'google-site-kit' ), $param ),
+			array( 'status' => 400 )
+		);
+	}
```
Comment (aaemnnosttv) [inline includes/Core/Notifications/Notifications.php]: @felixarntz I've thought about refactoring with something like this before since there is quite a bit of duplication for this kind of thing in the datapoint handlers. Ideally we would use core's `args` definitions and leverage the native validation the API already provides. The problem is that our data requests stuff all the parameters into one `data` parameter (except for a few outliers like module activation), and the REST API doesn't support nested validation except for collections. This may not be the best place to discuss this but just thought I'd raise it.

Comment (felixarntz) [reply includes/Core/Notifications/Notifications.php]: Yeah I thought about this initially too, but our support for batch requests would make use of that core feature challenging. I think for now we should keep this internal to the respective handlers (although adding utility methods, potentially even a validation class or trait for common validation features is certainly worth considering).

File: tests/phpunit/integration/Core/REST_API/REST_RoutesTest.php [hunk 4]
```diff
@@ -53,6 +53,8 @@ public function test_register() {
 			'/' . REST_Routes::REST_ROOT . '/core/site/data/developer-plugin',
 			'/' . REST_Routes::REST_ROOT . '/core/site/data/setup-tag',
 			'/' . REST_Routes::REST_ROOT . '/core/site/data/connection',
+			'/' . REST_Routes::REST_ROOT . '/core/site/data/notifications',
+			'/' . REST_Routes::REST_ROOT . '/core/site/data/mark-notification',
```
[the remark this answers is not in the document]
Comment (felixarntz) [reply tests/phpunit/integration/Core/REST_API/REST_RoutesTest.php]: Yeah that makes sense. Could you do that in a follow-up PR? If we can get it into this release, we can even just connect it to the same issue. I just don't want that work to block this PR.

**Comment by felixarntz** [comment]
Current appearance of the notification:
Screenshot 2020-02-25 at 14 19 08

**Comment by felixarntz** [comment]
@aaemnnosttv Could you maybe look into how we could support some basic inline HTML tags in the `content` returned? Would be useful to support at least a few whitelisted things such as `strong`, `em`, `br`, `a href`. It would be great if we could include that in this initial implementation, particularly because we would want to break the notification text into two lines with `br`.

File: includes/Core/Notifications/Notification.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+	/**
+	 * Prepares the JS representation of the Notification.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array
+	 */
+	public function prepare_for_js() {
+		return array(
+			'slug'           => $this->get_slug(),
```
Comment (felixarntz) [inline includes/Core/Notifications/Notification.php]: @aaemnnosttv That was not originally the plan, but maybe we should change this to `id` (just here, for JS)? "slug" is kind of the term only used in PHP, for consistency maybe `id` would be better for JS.

Comment (aaemnnosttv) [reply includes/Core/Notifications/Notification.php]: Sure. I assumed it should be included since we need it for identifying the notice in the click handlers. I'll update the key to `id`.

**Comment by kevinpdicarlo** [comment]
how come i am getting all theses threads??

Thank You,
Circle 22

Harmonize Humanity

File: includes/Core/Notifications/Notifications.php (hunk 3 above)
Comment (aaemnnosttv) [reply includes/Core/Notifications/Notifications.php]: > Yeah I thought about this initially too, but our support for batch requests would make use of that core feature challenging.

It would require that we change the way batch requests are handled to dispatch the individual REST requests (internally-w/o HTTP) in the batch handler rather than the way it is now but I think it's possible. Better to discuss this in a new issue though.

**Comment by aaemnnosttv** [comment]
@kevinpdicarlo - looks like you're [watching/following](https://github.com/google/site-kit-wp/watchers) the repository.

See here about updating what you get notified about:
https://help.github.com/en/github/receiving-notifications-about-activity-on-github/watching-and-unwatching-repositories

TL;DR, you're probably looking for "Not Watching" or "Releases Only" 😄 
image

**Comment by aaemnnosttv** [comment]
@felixarntz all of your feedback should be addressed now.

Regarding E2E, I updated the API mock to intercept requests to the proxy and return mock responses instead of overriding (and re-implementing) our core routes like we do in others. This makes it a bit more end-to-end and actually ensures that we are passing the necessary parameters to the endpoint as well. This is the better way to do it where possible I think.

As for the support for HTML in notification descriptions, I've added `dompurify` and allowed for the select elements and attributes you mentioned. One byproduct of this change is that we needed to separate the description into its own element (before it was sharing a `<p>`). This resulted in a few breaking changes to VRT but I think they are actually an improvement so I've gone ahead and approved them but you can see the full changes here.
[vrt-report.zip.txt](https://github.com/google/site-kit-wp/files/4254466/vrt-report.zip.txt)
_GH doesn't allow uploading zip files so just remove the `.txt` after downloading 😄_ 

image

image

**Comment by aaemnnosttv** [comment]
I had to bump the bundlesize limit slightly to allow for dompurify. I'll open a new issue about addressing the increasing size.

**Comment by tofumatt** [comment]
@googlebot I consent.

File: assets/js/components/notifications/dashboard-core-site-alerts.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+		this.state = {
+			notifications: false,
+		};
+	}
+
+	async componentDidMount() {
+		const notifications = await getNotifications().catch( ( e ) => {
+			// eslint-disable-next-line no-console
+			console.warn( 'Error caught while fetching notifications', JSON.stringify( e ) );
```
Comment (tofumatt) [inline assets/js/components/notifications/dashboard-core-site-alerts.js]: I often use `global.console` for these. It probably doesn't matter? I dunno!

Comment (tofumatt) [reply assets/js/components/notifications/dashboard-core-site-alerts.js]: Also: why stringify the error? `console` can output objects.

File: assets/js/components/notifications/dashboard-core-site-alerts.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+	constructor( props ) {
+		super( props );
+
+		this.state = {
+			notifications: false,
+		};
+	}
+
+	async componentDidMount() {
+		const notifications = await getNotifications().catch( ( e ) => {
```
Comment (tofumatt) [inline assets/js/components/notifications/dashboard-core-site-alerts.js]: For `async/await` code it's usually the norm to do try/catch blocks instead of `catch`. Any reason to prefer catch here?
```suggestion
		try {
			const notifications = await getNotifications();
			this.setState( { notifications } );
		} catch ( err ) {
			// eslint-disable-next-line no-console
			console.warn( 'Error caught while fetching notifications', JSON.stringify( err ) );
		}
```

I dunno what `notifications` get set to when this fails, but it seems off to set notifications if there was an error fetching them too.

File: assets/js/components/notifications/dashboard-core-site-alerts.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+	render() {
+		const { notifications } = this.state;
+
+		if ( ! Array.isArray( notifications ) ) {
+			return null;
+		}
+
+		return notifications.map(
+			( notification ) => <Notification
```
Comment (tofumatt) [inline assets/js/components/notifications/dashboard-core-site-alerts.js]: Just a nitpick, but I find this a bit tougher to read than:

```js
( notification ) => (
  <Notification
    {…props}
  />
)
```

File: assets/js/components/notifications/dashboard-core-site-alerts.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+				key={ notification.id }
+				id={ notification.id }
+				title={ notification.title || '' }
+				description={ notification.content || '' }
+				learnMoreURL={ notification.learnMoreURL || '' }
+				learnMoreLabel={ notification.learnMoreLabel || '' }
+				ctaLink={ notification.ctaURL || '' }
+				ctaLabel={ notification.ctaLabel || '' }
+				ctaTarget={ notification.ctaTarget || '' }
+				dismiss={ notification.dismissLabel || __( 'OK, Got it!', 'google-site-kit' ) }
```
Comment (tofumatt) [inline assets/js/components/notifications/dashboard-core-site-alerts.js]:
```suggestion
				dismiss={ notification.dismissLabel || __( 'OK, got it!', 'google-site-kit' ) }
```

The "G" here shouldn't be capitalised if not a new sentence.

File: includes/Core/Notifications/Notification.php [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+	 *     @type bool   $dismissible      Whether the notice should be dismissible. Default false.
+	 *     @type string $dismiss_label    Dismiss anchor text.
+	 * }
+	 */
+	public function __construct( $slug, array $args ) {
+		$this->slug = (string) $slug;
+		$this->args = array_merge(
+			array(
+				'title'            => '',
+				'content'          => '',
```
Comment (tofumatt) [inline includes/Core/Notifications/Notification.php]: Since these are always empty-string by default, any reason the component does `notification.content || ''`  when passing props? Seems unneeded. 🤷‍♂ Belt and braces I guess?

File: includes/Core/Notifications/Notifications.php (hunk 1 above)
Comment (tofumatt) [reply includes/Core/Notifications/Notifications.php]: I think it wants to `=>` to be aligned?

```php
'site_id'            => $credentials['oauth2_client_id'],
'site_secret'        => $credentials['oauth2_client_secret'],
'notification_id'    => $data['notificationID'],
'notification_state' => $data['notificationState'],
```

File: includes/Core/Notifications/Notifications.php (hunk 2 above)
Comment (tofumatt) [reply includes/Core/Notifications/Notifications.php]: A generic error message (like a 500, if the proxy's notification endpoint was down/erroring) seems wrong to display, because it's not very actionable/helpful to the user. I'd think an allow-list of errors to display would work if wanted, but otherwise it seems best to log them to the console for debugging/support. Otherwise hide 'em?

Comment (felixarntz) [reply includes/Core/Notifications/Notifications.php]: Yeah I think we should stick to only logging and don't render any user-facing messages. It should just silently error then.

File: includes/Core/Notifications/Notifications.php (hunk 1 above)
Comment (aaemnnosttv) [reply includes/Core/Notifications/Notifications.php]: It should want that, but it automatically fixes it as highlighted above (which should be wrong)

File: assets/js/components/notifications/dashboard-core-site-alerts.js (hunk 6 above)
Comment (aaemnnosttv) [reply assets/js/components/notifications/dashboard-core-site-alerts.js]: Yeah, I know but it doesn't work when logging from E2E tests.

https://travis-ci.com/google/site-kit-wp/jobs/291425794#L1104

I didn't mean to keep it like that though so I'll update it.

File: includes/Core/Notifications/Notification.php (hunk 10 above)
Comment (aaemnnosttv) [reply includes/Core/Notifications/Notification.php]: This ensures that the keys are always set on the PHP side.

**Comment by aaemnnosttv** [comment]
> Can you clarify why this is necessary? Wouldn't it be possible to use `Fragment dangerouslySetInnerHTML=...` instead? It would be great if we could avoid that change, so that it doesn't result in additional paragraphs just for the learn more link.

@felixarntz `Fragment`s don't support using `dangerouslySetInnerHTML` 😞 
See https://github.com/facebook/react/issues/12014

> It would be great if we could avoid that change, so that it doesn't result in additional paragraphs just for the learn more link.

I think the UI is a bit cleaner and more consistent with the separation myself, but if we wanted to keep them, we could wrap `description` and the learn more with `span`s instead and then wrap that with a `<p>` tag to let them render inline.

> These visual changes seem off to me:

@tofumatt They are _different_, but not unexpected. Like I said, I think they're an improvement but we can revert that if necessary.
See https://github.com/google/site-kit-wp/pull/1170#issuecomment-591315512

**Comment by felixarntz** [comment]
@aaemnnosttv 
> I think the UI is a bit cleaner and more consistent with the separation myself, but if we wanted to keep them, we could wrap `description` and the learn more with `span`s instead and then wrap that with a `<p>` tag to let them render inline.

Yeah I thought about that too, let's use `span`s then and maintain the original appearance.

File: assets/js/components/notifications/dashboard-core-site-alerts.js (hunk 9 above)
Comment (aaemnnosttv) [reply assets/js/components/notifications/dashboard-core-site-alerts.js]: There's a bunch of these in the codebase. I've updated them all so let's see how that goes...

File: assets/js/components/notifications/dashboard-core-site-alerts.js [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+					description={ notification.content || '' }
+					learnMoreURL={ notification.learnMoreURL || '' }
+					learnMoreLabel={ notification.learnMoreLabel || '' }
+					ctaLink={ notification.ctaURL || '' }
+					ctaLabel={ notification.ctaLabel || '' }
+					ctaTarget={ notification.ctaTarget || '' }
+					dismiss={ notification.dismissLabel || __( 'OK, Got it!', 'google-site-kit' ) }
+					isDismissable={ notification.dismissible }
+					onCTAClick={ async () => {
+						await acceptNotification( notification.id );
```
Comment (felixarntz) [inline assets/js/components/notifications/dashboard-core-site-alerts.js]: We need to make the notice dismiss here, also when the user accepts it. It could either be through manual logic in this callback here, or alternatively we could introduce some kind of `dismissOnCTA` flag prop in `Notification` to generally support that?

Comment (felixarntz) [reply assets/js/components/notifications/dashboard-core-site-alerts.js]: The thing is right now, it opens the link in a new tab (which is what should happen). The problem is then the notice still remains open in the other window. We need to make sure it dismissed on the JS side.

Comment (aaemnnosttv) [reply assets/js/components/notifications/dashboard-core-site-alerts.js]: I just pushed a small update to address this although it needs testing still.

Basically, if no `onCTAClick` handler is passed, the notification will dismiss on CTA click if `dismissible`.

If an `onCTAClick` handler _is passed_ then that callback is responsible for calling `dismiss` or not, which is passed to the callback as the second argument (first being the event). If the event isn't dismissible, the `dismiss` function is a no-op.

File: assets/js/components/notifications/notification.js [hunk 12]
```diff
@@ -74,6 +89,19 @@ class Notification extends Component {
...
+		e.persist();
+
+		const { isDismissable, onCTAClick } = this.props;
+		const dismiss = isDismissable ? this.dismiss : () => {};
+
+		if ( onCTAClick ) {
+			onCTAClick( e, dismiss );
+		} else {
+			dismiss( e );
+		}
```
Comment (felixarntz) [inline assets/js/components/notifications/notification.js]: Revisiting this, wouldn't it make more sense to make `handleCTAClick` `async` and then `await` `onCTAClick`, without passing `dismiss` to it? Then we would always call `dismiss` at the end. This would make it more similar to `handleDismiss` and avoid the extra parameter.

File: assets/js/components/notifications/notification.js [hunk 13]
```diff
@@ -55,14 +60,24 @@ class Notification extends Component {
 		}
 	}
 
-	handleDismiss( e ) {
-		const { isClosed } = this.state;
-		const card = this.cardRef.current;
-
+	async handleDismiss( e ) {
+		e.persist();
```
Comment (felixarntz) [inline assets/js/components/notifications/notification.js]: What's this for?

Comment (aaemnnosttv) [reply assets/js/components/notifications/notification.js]: Since these events are actually [`SyntheticEvents`](https://reactjs.org/docs/events.html#overview) from React, we need to persist them using this method to ensure they are not disposed of to make them available to any subsequent callbacks. We don't actually use the events right now, but if we're going to pass it, it should be persisted.

File: assets/js/components/notifications/notification.js (hunk 12 above)
Comment (aaemnnosttv) [reply assets/js/components/notifications/notification.js]: Sure we could do that. My initial approach here assumes that there might be cases where we don't want to dismiss on CTA click, even if the notification is dismissible, but that does sound like a bit of an edge case 😄 

This looks cleaner indeed though so assuming the above isn't a concern, let's go with it 👍

**Comment by aaemnnosttv** [comment]
@felixarntz I've simplified the `handleCTAClick` as you suggested and testing has been going well.
Ready for another pass 👍
